docs/cryptodoc/src/05_08_ml_dsa.rst

```diff
@@ -0,0 +1,443 @@
+      15. :math:`h = make\_hint(r_0 + ct_0, w_1)`
+      16. *Retry* if the Hamming weight of :math:`h > \omega`
+
+   4. :math:`\sigma = (\tilde{c}, z, h)` encoded using ``encode_signature``
+
+   **Notes:**
+
+   - This algorithm description assumes that the private signing key is expanded
+     into the internal representation already. Additionally, the expansion of
+     :math:`\hat{A}`, as well as the NTT for :math:`s_1`, :math:`s_2`, and
+     :math:`t_0` are done :srcref:`prior to the actual signing operation
+     <src/lib/pubkey/dilithium/dilithium_common/dilithium.cpp:131|Dilithium_Signature_Operation>`
+     to armortize the complexity of these operations across multiple consecutive
+     signature generations.
+   - Step 1: Botan 3.6.0 does not yet support the application-defined context
+     string as specified in [FIPS-204]_ Algorithm 2. See `GitHub #4376
+     <https://github.com/randombit/botan/issues/4376>`_.
+   - Step 3.12: We calculate :math:`r_0 = w_0 - cs_2`, this is equivalent to
+     :math:`r_0 = LowBits(w - cs_2)` as specified in [FIPS-204]_ Algorithm 7.
+   - Step 3.15: We generate the hint from :math:`w_0 - cs_2 + ct_0` this is equivalent
+     to the specification's :math:`w - cs_2 + ct_0`.
+
+.. _pubkey/ml_dsa/verification:
+
+Signature Verification
+----------------------
+
+Signature verification as specified in [FIPS-204]_ Algorithms 3 and 8 are
+implemented in :srcref:`Dilithium_Verification_Operation::is_valid_signature
+<src/lib/pubkey/dilithium/dilithium_common/dilithium.cpp:269|is_valid_signature>`
+with the preparation of the message representative :math:`\mu` being done in
+:srcref:`DilithiumMessageHash
+<src/lib/pubkey/dilithium/dilithium_common/dilithium_symmetric_primitives.h:31|DilithiumMessageHash>`.
+
+.. admonition:: Dilithium_Verification_Operation::is_valid_signature
+
+   **Input:**
+
+   - ``pk``: public verification key, with :math:`\hat{A}` and :math:`\hat{t_1'} = NTT(t_1 \cdot 2^{d})` in NTT domain
+   - ``M``: message to be verified
+   - ``signature``: the signature to be verified
+
+   **Output:**
+
+   - ``ok``: boolean value whether or not the signature is valid
+
+   **Steps:**
+
+   1. Calculate the message representative :math:`\mu = H(H(pk) \| 0x00 \| 0x00 \| M)` (see :srcref:`here <src/lib/pubkey/dilithium/ml_dsa/ml_dsa_impl.h:36|start>`)
+   2. Decode the signature into :math:`(\tilde{c}, z, h)` using ``decode_signature``
+   3. *Abort with "not valid"* if the Hamming weight of :math:`h > \omega`
+   4. *Abort with "not valid"* if :math:`\|z\|_{\infty} \geq \gamma_1 - \beta` (see ``infinity_norm_within_bound``)
+   5. :math:`\hat{c} = NTT(sample\_in\_ball(\tilde{c}))`
+   6. :math:`w'_{approx} = \hat{A} \cdot NTT(z) - \hat{c} \cdot \hat{t_1'}`
+   7. :math:`w_1' = use\_hint(w'_{approx}, h)`
+   8. :math:`\tilde{c'} = H(\mu, w_1')` (:math:`w_1'` is encoded using ``encode_commitment``)
+   9. Iff :math:`\tilde{c} = \tilde{c'}` *return "valid"*, else *"not valid"*
```

Comment:
```suggestion
   8. :math:`\tilde{c}' = H(\mu, w_1')` (:math:`w_1'` is encoded using ``encode_commitment``)
   9. If :math:`\tilde{c} = \tilde{c}'` *return "valid"*, else *"not valid"*
```
Looks more like in the spec. Also the _iff_ is not needed, I think, because of the _else_.